README.md
# node-fronius-solar

[![Greenkeeper badge](https://badges.greenkeeper.io/mwittig/node-fronius-solar.svg)](https://greenkeeper.io/) [![Build Status](https://travis-ci.org/mwittig/node-fronius-solar.svg?branch=master)](https://travis-ci.org/mwittig/node-fronius-solar)

Access PV live logs using the Fronius Solar API V0 and V1.

## Usage Example

    var fronius = require('node-fronius-solar'),
        util = require('util'),
        options = {
            host: 'localhost',
            port: 8001,
            deviceId: 1
            version: 0
        },
        x = Date.now();
    
    // This is the Solar API V0 call which should work with Fronius Datalogger Web v2.0.4.1 (and higher) and
    // Fronius Data Manager v3.0.3-1 (and higher)
    fronius.GetInverterRealtimeData(options).then(function (json) {
        console.log(util.inspect(json, { depth: 4, colors : true }));
        console.log(Date.now() - x, "milliseconds elapsed")
    }).catch(function(e) {console.log(e)});
    
    // This is the Solar API V1
    options.version = 1;
    fronius.GetInverterRealtimeData(options).then(function (json) {
        console.log(util.inspect(json, { depth: 4, colors : true }));
        console.log(Date.now() - x, "milliseconds elapsed")
    }).catch(function(e) {console.log(e)});
    
    // GetComponentsData is provided to use an undocumented API service of the Fronius Data Manager
    // provided with the Symo inverters. See https://forum.fhem.de/index.php/topic,24614.msg214011.html#msg214011
    // In contrast to the regular Solar API Calls which do not require authentication, you may need to provide
    // username and password properties to the options dictionary (if authentication has been enabled 
    // for Fronius Data Manager.
    fronius.GetComponentsData(options).then(function (json) {
        console.log(util.inspect(json, { depth: 4, colors : true }));
        console.log(Date.now() - x, "milliseconds elapsed")
    }).catch(function(e) {console.log(e)});
    
    // This is a Solar API V1 call which should work with Fronius Data Manager v3.4.2-1 (and higher)
    fronius.GetPowerFlowRealtimeDataData(options).then(function (json) {
        console.log(util.inspect(json, { depth: 4, colors : true }));
        console.log(Date.now() - x, "milliseconds elapsed")
    }).catch(function(e) {console.log(e)});

## Release History

See [Release History](https://github.com/mwittig/node-fronius-solar/blob/master/HISTORY.md).

## License

Copyright (c) 2016, Marcus Wittig and contributors. All rights reserved.

[MIT License](https://github.com/mwittig/node-fronius-solar/blob/master/LICENSE).


HISTORY.md
# Release History

* 20150512, V0.0.1
    * Initial Version

* 20150514, V0.0.2
    * Improved error handling
    * Added support for HTTPS, added rejectUnauthorized: false to allow self-signed server certs. Should be set to true
      if server has a certificate signed from a trusted CA

* 20150518, V0.0.3
    * Improved error handling
    
* 20160305, V0.0.4
    * Updated dependencies
    * Replaced deprecated usage of Promise.settle() function
    * Added travis build descriptor
    
* 20160318, V0.0.5
    * Added GetComponentsData() for using an undocumented API service provided by the data logger 
      of the Symo inverters. Issue #1
    * Moved server.js to 'test' directory as a starting pointed for automating testing
    * Updated dependencies
    
* 20160330, V0.0.6
    * Added HTTP-Digest Authentication as required for GetComponentsData if authentication has been enabled 
      for Fronius Data Manager
    * Added GetPowerFlowRealtimeDataData() Solar API V1 service call
    * Moved release history to separate file
    * Added license info to README

* 20160601, V0.0.7
    * Added basic support for V0 GetInverterRealtimeData

* 20160601, V0.0.8
    * Bug fix for V0 GetInverterRealtimeData
    
* 20160921, V0.0.9
    * Updated dependencies
    
* 20170929, V0.0.10
    * Updated dependencies
    * Added Greenkeeper badge

test/server.js
"use strict";

var http = require("http"),
    url = require("url"),
    path = require("path"),
    auth = require('http-auth'),
    crypto = require('crypto'),
    port = process.argv[2] || 8001,
    resourcePaths = {
        REALTIME_DATA_V0: '/solar_api/GetInverterRealtimeData.cgi',
        REALTIME_DATA: '/solar_api/v1/GetInverterRealtimeData.cgi',
        COMPONENTS: '/components/5/0/?print=names',
        POWER_FLOW_REALTIME_DATA: '/solar_api/v1/GetPowerFlowRealtimeData.fcgi'
    },
    realm = "PV Data Logger",
    useDigestAuth = true;
//solar_api/GetInverterRealtimeData.cgi?Scope=Device&DeviceIndex=1&DataCollection=CommonInverterData
function normalize(path) {
    // do some minimal normalization of the URL path, i.e. remove double slashes
    return path.replace(/\/{2,}/, '/')
}

var requestListener = function(request, response) {

    var requestUrl = url.parse(request.url);
    console.log("Request URL path", requestUrl.path);

    if (normalize(requestUrl.pathname) === resourcePaths.REALTIME_DATA_V0) {
        console.log("200 OK");
        response.writeHead(200, {
            "Content-Type": "application/json; charset=ISO-8859-1"
        });
        response.end(JSON.stringify(
            {
                "Head" : {
                    "RequestArguments" : {
                        "DataCollection" : "CommonInverterData",
                        "DeviceClass" : "Inverter",
                        "DeviceIndex" : "1",
                        "Scope" : "Device"
                    },
                    "Status" : {
                        "Code" : 0,
                        "Reason" : "",
                        "UserMessage" : ""
                    },
                    "Timestamp" : "2016-06-01T15:31:47+02:00"
                },
                "Body" : {
                    "Data" : {
                        "DAY_ENERGY" : {
                            "Value" : 10000,
                            "Unit" : "Wh"
                        },
                        "FAC" : {
                            "Value" : 49.97,
                            "Unit" : "Hz"
                        },
                        "IAC" : {
                            "Value" : 8.42,
                            "Unit" : "A"
                        },
                        "IDC" : {
                            "Value" : 8.48,
                            "Unit" : "A"
                        },
                        "PAC" : {
                            "Value" : 2021,
                            "Unit" : "W"
                        },
                        "TOTAL_ENERGY" : {
                            "Value" : 70334000,
                            "Unit" : "Wh"
                        },
                        "UAC" : {
                            "Value" : 240,
                            "Unit" : "V"
                        },
                        "UDC" : {
                            "Value" : 253,
                            "Unit" : "V"
                        },
                        "YEAR_ENERGY" : {
                            "Value" : 3746000,
                            "Unit" : "Wh"
                        },
                        "DeviceStatus" : {
                            "StatusCode" : 7,
                            "MgmtTimerRemainingTime" : -1,
                            "ErrorCode" : 0,
                            "LEDColor" : 2,
                            "LEDState" : 0,
                            "StateToReset" : false
                        }
                    }
                }
            }
        ));
        return;
    }
    else if (normalize(requestUrl.pathname) === resourcePaths.REALTIME_DATA) {
        console.log("200 OK");
        response.writeHead(200, {
            "Content-Type": "application/json; charset=ISO-8859-1"
        });
        response.end(JSON.stringify(
            {
                "Head": {
                    "RequestArguments": {
                        "DataCollection": "CommonInverterData",
                        "DeviceClass": "Inverter",
                        "DeviceId": "1",
                        "Scope": "Device"
                    },
                    "Status": {
                        "Code": 0,
                        "Reason": "",
                        "UserMessage": ""
                    },
                    "Timestamp": "2015-05-07T16:55:31+02:00"
                },
                "Body": {
                    "Data": {
                        "DAY_ENERGY": {
                            "Value": 26000,
                            "Unit": "Wh"
                        },
                        "FAC": {
                            "Value": 49.97,
                            "Unit": "Hz"
                        },
                        "IAC": {
                            "Value": 5.4,
                            "Unit": "A"
                        },
                        "IDC": {
                            "Value": 3.78,
                            "Unit": "A"
                        },
                        "PAC": {
                            "Value": 1251,
                            "Unit": "W"
                        },
                        "TOTAL_ENERGY": {
                            "Value": 2376359,
                            "Unit": "Wh"
                        },
                        "UAC": {
                            "Value": 228,
                            "Unit": "V"
                        },
                        "UDC": {
                            "Value": 428.7,
                            "Unit": "V"
                        },
                        "YEAR_ENERGY": {
                            "Value": 2005598,
                            "Unit": "Wh"
                        },
                        "DeviceStatus": {
                            "StatusCode": 7,
                            "MgmtTimerRemainingTime": -1,
                            "ErrorCode": 0,
                            "LEDColor": 2,
                            "LEDState": 0,
                            "StateToReset": false
                        }
                    }
                }
            }
        ));
        return;
    }
    else if (normalize(requestUrl.path) === resourcePaths.COMPONENTS) {
        console.log("200 OK");
        response.writeHead(200, {
            "Content-Type": "application/json; charset=ISO-8859-1"
        });
        response.end(JSON.stringify(
            {
                "Head": {
                    "RequestArguments": {
                        "resource": "5\/0\/",
                        "print": "names"
                    },
                    "Status": {
                        "Code": 0,
                        "Reason": "",
                        "UserMessage": "",
                        "ErrorDetail": {
                            "Nodes": []
                        }
                    },
                    "Timestamp": "2016-02-25T11:23:46+01:00"
                },
                "Body": {
                    "Data": {
                        "TimeStamp": {
                            "value": 1456395825,
                            "unit": "sec"
                        },
                        "Enable": {
                            "value": 1,
                            "unit": "1"
                        },
                        "Visible": {
                            "value": 1,
                            "unit": "1"
                        },
                        "Power_P_Generate": {
                            "value": 629,
                            "unit": "W"
                        },
                        "Power_P_Load": {
                            "value": -2883.25,
                            "unit": "W"
                        },
                        "Power_P_Grid": {
                            "value": 2254.25,
                            "unit": "W"
                        },
                        "Power_Akku_Sum": {
                            "value": null,
                            "unit": "W"
                        },
                        "Power_PV_Sum": {
                            "value": 629,
                            "unit": "W"
                        },
                        "Relative_Current_SelfConsumption": {
                            "value": 100,
                            "unit": "%"
                        },
                        "Relative_Current_Autonomy": {
                            "value": 21.815659,
                            "unit": "%"
                        },
                        "Power_P_SelfConsumption": {
                            "value": -629,
                            "unit": "W"
                        }
                    }
                }
            }
        ));
        return;
    }
    else if (normalize(requestUrl.path) === resourcePaths.POWER_FLOW_REALTIME_DATA) {
        console.log("200 OK");
        response.writeHead(200, {
            "Content-Type": "application/json; charset=ISO-8859-1"
        });
        response.end(JSON.stringify(
            {
                "Head" : {
                    "RequestArguments" : {},
                    "Status" : {
                        "Code" : 0,
                        "Reason" : "",
                        "UserMessage" : ""
                    },
                    "Timestamp" : "2016-03-25T15:19:21+01:00"
                },
                "Body" : {
                    "Data" : {
                        "Site" : {
                            "Mode" : "produce-only",
                            "P_Grid" : null,
                            "P_Load" : null,
                            "P_Akku" : null,
                            "P_PV" : 291,
                            "E_Day" : 7653,
                            "E_Year" : 841228,
                            "E_Total" : 7978902
                        },
                        "Inverters" : {
                            "1" : {
                                "DT" : 110,
                                "P" : 291
                            }
                        }
                    }
                }
            }
        ));
        return;
    }
    else {
        console.log("404 Not Found");
        response.writeHead(404, {"Content-Type": "text/plain"});
        response.write("404 Not Found\n");
        response.end();
        return;
    }
}

if (useDigestAuth) {
    var digest = auth.digest({
            realm: realm
        }, function (username, callback) { // Expecting md5(username:realm:password) in callback.
            if (username === "admin") {
                var hash = crypto.createHash('md5');
                hash.update("admin:" + realm + ":admin");
                callback(hash.digest('hex'));
            } else {
                callback();
            }
        }
    );
    http.createServer(digest, requestListener).listen(parseInt(port, 10));
}
else {
    http.createServer(requestListener).listen(parseInt(port, 10));
}

console.log("Server running at\n  => http://localhost:" + parseInt(port, 10) + "/\nCTRL + C to shutdown");

index.js
var util = require('util'),
    http = require('http'),
    https = require('https'),
    _ = require('lodash'),
    Promise = require('bluebird'),
    lastRequest = Promise.resolve(),
    debug = process.env.hasOwnProperty('FRONIUS_DEBUG') ? consoleDebug : function () {
    },
    DEFAULT_DEVICE_ID = 1;

//
// Private Helper Functions
//

function consoleDebug() {
    console.log.apply(this, arguments)
}

function settlePromise(aPromise) {
    return aPromise.reflect();
}

function getRequest(options, path) {
    var id = options.deviceId || DEFAULT_DEVICE_ID;
    var urlPath = path;
    if (! urlPath) {
        if (options.version === 0) {
            urlPath = '/solar_api/GetInverterRealtimeData.cgi?Scope=Device&DeviceIndex=';
        }
        else {
            urlPath = '/solar_api/v1/GetInverterRealtimeData.cgi?Scope=Device&DeviceId=';
        }
        urlPath+= id + '&DataCollection=CommonInverterData'
    }
    var requestOptions = _.assign({
        deviceId: DEFAULT_DEVICE_ID,
        timeout: 20000,
        rejectUnauthorized: false,
        protocol: 'http:',
        path: urlPath,
        headers: {},
        method: 'GET'
    }, options);
    var timeoutOccurred = false;

    requestOptions.headers['Host'] = requestOptions.host;
    //if (!_.isUndefined(requestOptions.username) && !_.isUndefined(requestOptions.password)) {
    //    requestOptions.headers['Authorization'] =
    //        "Basic " + new Buffer(requestOptions.username + ":" + requestOptions.password).toString("base64");
    //}

    debug('REQUEST OPTIONS: ' + JSON.stringify(requestOptions));

    return new Promise(function (resolve, reject) {
        var data = "";
        var proto = (requestOptions.protocol == 'https:') ? https : http;
        if  (requestOptions.username != null) {
            proto = require('http-digest-client')(
                requestOptions.username,
                requestOptions.password,
                (requestOptions.protocol == 'https:')
            );
        }
        var getReq = proto.request(requestOptions, function (response) {
                debug('STATUS: ' + response.statusCode);
                debug('HEADERS: ' + JSON.stringify(response.headers));

                var error;
                if (response.statusCode >= 300) {
                    if (response.statusCode === 401) {
                        error = new Error("Unauthorized: check username/password");
                    }
                    else {
                        error = new Error("Request failed. HTTP Status Code: " + response.statusCode);
                    }
                    debug('ERROR:' + 'Host ' + requestOptions.host + ' ' + error);
                    return reject(error);
                }

                response.setEncoding('utf8');
                response.on('data', function (result) {
                    debug("DATA CHUNK", result);
                    data += result;
                });
                response.on('end', function () {
                    debug("END");
                    try {
                        var json = JSON.parse(data);
                        if (_.has(json, 'Head') && _.has(json, 'Body')) {
                            return resolve(json);
                        }
                        else {
                            var error = new Error("Invalid response body format: Head and Body expected");
                            debug('ERROR:' + 'Host ' + requestOptions.host + ' ' + error);
                            return reject(error);
                        }
                    }
                    catch (e) {
                        var error = new Error("Invalid response body: " + e.toString());
                        debug('ERROR:' + 'Host ' + requestOptions.host + ' ' + error);
                        return reject(error);
                    }
                });
            }).on('error', function (error) {
                if (timeoutOccurred) {
                    error = new Error("Request timeout occurred - request aborted");
                }
                debug('ERROR:' + 'Host ' + requestOptions.host + ' ' + error);
                getReq.abort();
                return reject(error);
            }).on('timeout', function () {
                timeoutOccurred = true;
                getReq.abort();
            });
        getReq.setTimeout(requestOptions.timeout);
        getReq.end();
    });
}

function checkRequiredProperties(options, requiredPropsArray) {
    for (var i = 0; i < requiredPropsArray.length; ++i) {
        if (_.isUndefined(options[requiredPropsArray[i]])) {
            return Promise.reject("Request options lacks required property=" + requiredPropsArray[i]);
        }
    }
    return Promise.resolve()
}

//
// Public Functions
//


module.exports.GetInverterRealtimeData = function (options) {
    var opts = _.clone(options);
    return checkRequiredProperties(opts, ['host', 'deviceId']).then(function () {
        return lastRequest = settlePromise(lastRequest).then(function () {
            return getRequest(opts).then(function (json) {
                return Promise.resolve(json);
            })
        })
    })
};

// GetComponentsData is provided to use an undocumented API service provided by the data logger
// of the Symo inverters. See https://forum.fhem.de/index.php/topic,24614.msg214011.html#msg214011
module.exports.GetComponentsData = function (options) {
    var opts = _.clone(options);
    return checkRequiredProperties(opts, ['host']).then(function () {
        return lastRequest = settlePromise(lastRequest).then(function () {
            return getRequest(opts, '/components/5/0/?print=names').then(function (json) {
                return Promise.resolve(json);
            })
        })
    })
};

module.exports.GetPowerFlowRealtimeDataData = function (options) {
    var opts = _.clone(options);
    return checkRequiredProperties(opts, ['host']).then(function () {
        return lastRequest = settlePromise(lastRequest).then(function () {
            return getRequest(opts, '/solar_api/v1/GetPowerFlowRealtimeData.fcgi').then(function (json) {
                return Promise.resolve(json);
            })
        })
    })
};


example.js
var fronius = require('./index'),
    util = require('util'),
    options = {
        username: "admin",
        password: "admin",
        host: 'localhost',
        port: 8001,
        deviceId: 1,
        version: 0
    },
    x = Date.now();

// This is the Solar API V0 call which should work with Fronius Datalogger Web v2.0.4.1 (and higher) and
// Fronius Data Manager v3.0.3-1 (and higher)
fronius.GetInverterRealtimeData(options).then(function (json) {
    console.log(util.inspect(json, { depth: 4, colors : true }));
    console.log(Date.now() - x, "milliseconds elapsed")
}).catch(function(e) {console.log(e)});

// This is the Solar API V1
options.version = 1;
fronius.GetInverterRealtimeData(options).then(function (json) {
    console.log(util.inspect(json, { depth: 4, colors : true }));
    console.log(Date.now() - x, "milliseconds elapsed")
}).catch(function(e) {console.log(e)});

// GetComponentsData is provided to use an undocumented API service of the Fronius Data Manager
// provided with the Symo inverters. See https://forum.fhem.de/index.php/topic,24614.msg214011.html#msg214011
// In contrast to the regular Solar API Calls which do not require authentication, you may need to provide
// username and password properties to the options dictionary (if authentication has been enabled
// for Fronius Data Manager.
fronius.GetComponentsData(options).then(function (json) {
    console.log(util.inspect(json, { depth: 4, colors : true }));
    console.log(Date.now() - x, "milliseconds elapsed")
}).catch(function(e) {console.log(e)});

// This is a Solar API V1 call which should work with Fronius Data Manager v3.4.2-1 (and higher)
fronius.GetPowerFlowRealtimeDataData(options).then(function (json) {
    console.log(util.inspect(json, { depth: 4, colors : true }));
    console.log(Date.now() - x, "milliseconds elapsed")
}).catch(function(e) {console.log(e)});

server.js
var http = require("http"),
    url = require("url"),
    path = require("path"),
    port = process.argv[2] || 8001;

http.createServer(function(request, response) {

    var requestUrl = url.parse(request.url);
    console.log("Request URL path", requestUrl.path);
    if (requestUrl.pathname === '/solar_api/v1/GetInverterRealtimeData.cgi') {
        response.writeHead(200, {
            "Content-Type": "application/json; charset=ISO-8859-1"
        });
        response.end(JSON.stringify(
            {
                "Head": {
                    "RequestArguments": {
                        "DataCollection": "CommonInverterData",
                        "DeviceClass": "Inverter",
                        "DeviceId": "1",
                        "Scope": "Device"
                    },
                    "Status": {
                        "Code": 0,
                        "Reason": "",
                        "UserMessage": ""
                    },
                    "Timestamp": "2015-05-07T16:55:31+02:00"
                },
                "Body": {
                    "Data": {
                        "DAY_ENERGY": {
                            "Value": 26000,
                            "Unit": "Wh"
                        },
                        "FAC": {
                            "Value": 49.97,
                            "Unit": "Hz"
                        },
                        "IAC": {
                            "Value": 5.4,
                            "Unit": "A"
                        },
                        "IDC": {
                            "Value": 3.78,
                            "Unit": "A"
                        },
                        "PAC": {
                            "Value": 1251,
                            "Unit": "W"
                        },
                        "TOTAL_ENERGY": {
                            "Value": 2376359,
                            "Unit": "Wh"
                        },
                        "UAC": {
                            "Value": 228,
                            "Unit": "V"
                        },
                        "UDC": {
                            "Value": 428.7,
                            "Unit": "V"
                        },
                        "YEAR_ENERGY": {
                            "Value": 2005598,
                            "Unit": "Wh"
                        },
                        "DeviceStatus": {
                            "StatusCode": 7,
                            "MgmtTimerRemainingTime": -1,
                            "ErrorCode": 0,
                            "LEDColor": 2,
                            "LEDState": 0,
                            "StateToReset": false
                        }
                    }
                }
            }
        ));
        return;
    }
    else if (requestUrl.path === '/components/5/0/?print=names') {
        response.writeHead(200, {
            "Content-Type": "application/json; charset=ISO-8859-1"
        });
        response.end(JSON.stringify(
            {
                "Head": {
                    "RequestArguments": {
                        "resource": "5\/0\/",
                        "print": "names"
                    },
                    "Status": {
                        "Code": 0,
                        "Reason": "",
                        "UserMessage": "",
                        "ErrorDetail": {
                            "Nodes": []
                        }
                    },
                    "Timestamp": "2016-02-25T11:23:46+01:00"
                },
                "Body": {
                    "Data": {
                        "TimeStamp": {
                            "value": 1456395825,
                            "unit": "sec"
                        },
                        "Enable": {
                            "value": 1,
                            "unit": "1"
                        },
                        "Visible": {
                            "value": 1,
                            "unit": "1"
                        },
                        "Power_P_Generate": {
                            "value": 629,
                            "unit": "W"
                        },
                        "Power_P_Load": {
                            "value": -2883.25,
                            "unit": "W"
                        },
                        "Power_P_Grid": {
                            "value": 2254.25,
                            "unit": "W"
                        },
                        "Power_Akku_Sum": {
                            "value": null,
                            "unit": "W"
                        },
                        "Power_PV_Sum": {
                            "value": 629,
                            "unit": "W"
                        },
                        "Relative_Current_SelfConsumption": {
                            "value": 100,
                            "unit": "%"
                        },
                        "Relative_Current_Autonomy": {
                            "value": 21.815659,
                            "unit": "%"
                        },
                        "Power_P_SelfConsumption": {
                            "value": -629,
                            "unit": "W"
                        }
                    }
                }
            }
        ));
        return;
    }
    else {
        response.writeHead(404, {"Content-Type": "text/plain"});
        response.write("404 Not Found\n");
        response.end();
        return;
    }
}).listen(parseInt(port, 10));

console.log("Server running at\n  => http://localhost:" + parseInt(port, 10) + "/\nCTRL + C to shutdown");